### PR TITLE
Module 5. Routing

### PR DESCRIPTION
**Pull request by VikaScherban**
- Add params to url: search, genre, sort
- Add param to url with movie id
- Load page with getting values from url
- Covered all url change with e2e tests
- Add all unit tests

### CODE REVIEW

File: src/components/MovieListPage.tsx [hunk 1]
```diff
@@ -1,46 +1,23 @@
...
-        search: '',
-        sortBy: SortByOptions.releaseDate.value,
-        activeGenre: '',
-        selectedMovie: null,
-    });
     const [movieDialogData, setMovieDialogData] = useState<MovieDialogState>({
         isOpen: false,
         title: '',
         movie: null,
     });
```
Comment (DmitriyDubravin) [inline src/components/MovieListPage.tsx:14]: If you not move this logic out to a custom hook to be reusable, there is no need to use 'isOpen' and 'title'. Just save here only a movie. isOpen === !!movie; title can be a static string

File: src/components/content/filter-line/FilterLine.tsx [hunk 2]
```diff
@@ -1,19 +1,33 @@
...
 import GenreSelect from "./genre-select/GenreSelect";
 import SortControl from "./sort-control/SortControl";
-import {FilterData} from "../../../models/filter";
 import {GenreTitle} from "../../../constants/genres-const";
 import React from "react";
+import useMultipleSearchParams from "../../../hooks/UseMultipleSearchParams";
+
+function FilterLine(): React.JSX.Element {
+    const { updateQueryParams } = useMultipleSearchParams();
+    const urlSearchParams = new URLSearchParams(location.search);
```
Comment (DmitriyDubravin) [inline src/components/content/filter-line/FilterLine.tsx:10]: Where is 'location' come from?

File: src/components/dialogs/movie-dialog/MovieDialog.tsx [hunk 3]
```diff
@@ -2,12 +2,20 @@ import BaseDialog from "../base-dialog/BaseDialog";
...
+import {Portal} from "react-portal";
 
-function MovieDialog({onClose, onSubmitChanges, title, movie}: MovieDialogData): React.JSX.Element {
+function MovieDialog({isOpen, onClose, onSubmitChanges, title, movie}: MovieDialogData): React.JSX.Element {
     return (
-        <BaseDialog title={title} onClose={onClose}>
-            <MovieForm onClose={onClose} movie={movie} onSubmitChanges={onSubmitChanges}/>
-        </BaseDialog>
+        <>
+            {isOpen &&
```
Comment (DmitriyDubravin) [inline src/components/dialogs/movie-dialog/MovieDialog.tsx:10]: You could return null instead. Just write a simple if statement before the main 'return'.
if (!isOpen) { retrun null; }

File: src/components/header/header-with-search/HeaderWithSearch.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import SearchForm from "../search-form/SearchForm";
+import MovieDialog from "../../dialogs/movie-dialog/MovieDialog";
+import {Movie, MovieDialogState} from "../../../models/movies";
+
+function HeaderWithSearch(): React.JSX.Element {
+    const [movieDialogData, setMovieDialogData] = useState<MovieDialogState>({
+        isOpen: false,
+        title: '',
+        movie: null,
+    });
```
Comment (DmitriyDubravin) [inline src/components/header/header-with-search/HeaderWithSearch.tsx:12]: Same here. No need for a complex state object. Just save the movie an it should be enough.

File: src/components/content/filter-line/FilterLine.tsx [hunk 5]
```diff
@@ -1,19 +1,33 @@
...
+    const {genre, sort} = Object.fromEntries(urlSearchParams.entries());
```
Comment (DmitriyDubravin) [inline src/components/content/filter-line/FilterLine.tsx:11]: Can you move this and 'new URLSearchParams' functionalities into 'useMulitpleSearchParams' hook?

File: src/components/header/movie-details/MovieDetails.tsx [hunk 6]
```diff
@@ -1,25 +1,35 @@
...
+    const { updateQueryParams } = useMultipleSearchParams();
     const convertToHours = (totalMinutes: number) => {
         const hours = Math.floor(totalMinutes / 60);
         const minutes = totalMinutes % 60;
 
         return `${hours}h ${minutes}min`;
     }
 
+    const goBack = () => {
+        updateQueryParams({}, '/');
```
Comment (DmitriyDubravin) [inline src/components/header/movie-details/MovieDetails.tsx:22]: You could return 'navigate' from your hook. And just use it here, as it should be used

File: src/components/header/search-form/SearchForm.tsx [hunk 7]
```diff
@@ -1,34 +1,33 @@
...
 import React, {useState} from "react";
+import useMultipleSearchParams from "../../../hooks/UseMultipleSearchParams";
 
-function SearchForm({ onSearchChanged, initialQuery }: SearchData): React.JSX.Element {
-    const [formData, setFormData] = useState({
-        query: initialQuery,
-    });
+function SearchForm(): React.JSX.Element {
+    const { updateQueryParams } = useMultipleSearchParams();
+    const urlSearchParams = new URLSearchParams(location.search);
```
Comment (DmitriyDubravin) [inline src/components/header/search-form/SearchForm.tsx:7]: location.search?

File: src/hooks/UseMultipleSearchParams.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            searchParams.set(key, params[key]);
+        });
+
+        navigate({
+            pathname: pathname || location.pathname,
+            search: `?${searchParams.toString()}`,
+        });
+    };
+
+    return { updateQueryParams };
```
Comment (DmitriyDubravin) [inline src/hooks/UseMultipleSearchParams.ts:35]: Pass 'navigate' here too. You'll gonna need it.

**Review by DmitriyDubravin — APPROVED** [review]

File: src/components/content/filter-line/FilterLine.tsx (hunk 5 above)
Comment (VikaScherban) [reply src/components/content/filter-line/FilterLine.tsx:11]: Sorry, I don't understand why to move this part responsible for reading url params into hook which is responsible for saving url params

**Pull request merged**